サンプルプロジェクト(example/aws_ec2)をRHEL7に対応させる

xirc opened the pull request:
Closes https://github.com/lerna-stack/lerna-terraform/issues/10

## 変更点

- モックサーバを CentOS8 と RHEL7 の両方で動作できるようにします
  Docker Engine のインストールを CentOS8 と RHEL7 の両方に対応させることは難しそうでした。  
  モックサーバは Node.js で動作しており、Node.js を CentOS8 と RHEL7 の両方にインストールすることは簡単でした。  
  Docker Engine をインストールする必要はないため、モックサーバを動作させるために Node.js を直接使用します。
- Kamon 1.x のホスト名解決でエラーにならないように変更します
  RHEL7 では、Kamon 1.x のホスト名解決がエラーになりました。  
  ホスト名の自動解決を使わず、明示的に指定するようにします。

## 動作確認
CentOS8 と RHEL7 の AMI それぞれを使った場合について、
[動作確認](https://github.com/lerna-stack/lerna-terraform/blob/main/examples/aws_ec2/README.md#%E5%8B%95%E4%BD%9C%E7%A2%BA%E8%AA%8D) を実施して、期待通りのレスポンスが返ってくることを確認しました。

Commit:
Mock server works on both CentoOS8 and RHEL7

Use npm instead of docker to run mock_server.
It may be an easy way to run the server on both CentOS8 and RHEL7.

Commit:
Reconcile an application configuration to work on RHEL7

Set a hostname config explicitly for Kamon.
It cannot be resolved automatically on Lerna Stack with RHEL7.

```
lerna-sample-payment-app[30861]: Caused by: java.net.UnknownHostException: ip-10-0-1-51.ap-northeast-1.compute.internal: ip-10-0-1-51.ap-northeast-1.compute.internal: Name or service not known
lerna-sample-payment-app[30861]: at java.net.InetAddress.getLocalHost(InetAddress.java:1506)
lerna-sample-payment-app[30861]: at kamon.Environment$.$anonfun$fromConfig$2(Environment.scala:37)
lerna-sample-payment-app[30861]: at kamon.Environment$.readValueOrGenerate(Environment.scala:49)
lerna-sample-payment-app[30861]: at kamon.Environment$.fromConfig(Environment.scala:37)
lerna-sample-payment-app[30861]: at kamon.Kamon$.<init>(Kamon.scala:36)
lerna-sample-payment-app[30861]: at kamon.Kamon$.<clinit>(Kamon.scala)
lerna-sample-payment-app[30861]: ... 104 more
lerna-sample-payment-app[30861]: Caused by: java.net.UnknownHostException: ip-10-0-1-51.ap-northeast-1.compute.internal: Name or service not known
lerna-sample-payment-app[30861]: at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
lerna-sample-payment-app[30861]: at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:929)
lerna-sample-payment-app[30861]: at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1324)
lerna-sample-payment-app[30861]: at java.net.InetAddress.getLocalHost(InetAddress.java:1501)
lerna-sample-payment-app[30861]: ... 109 more
```

xirc commented on examples/aws_ec2/facility-mock-server.tf:
```diff
@@ -37,88 +31,43 @@ resource "null_resource" "setup_for_mock_server" {
-
-  EOF
+Type = simple
+WorkingDirectory = ${local.mock-server-docker-filepath}
```
`WorkingDirectory` は次のページにドキュメントがあります。
https://www.freedesktop.org/software/systemd/man/systemd.exec.html#WorkingDirectory=

negokaz commented on examples/aws_ec2/facility-mock-server.tf:
```diff
@@ -37,88 +31,43 @@ resource "null_resource" "setup_for_mock_server" {
+WorkingDirectory = ${local.mock-server-docker-filepath}
+ExecStartPre = /usr/bin/npm install
+ExecStart = /usr/bin/npm start -- --host '0.0.0.0' --port 8083
+ExecStop = /usr/bin/npm stop
```
@xirc `ExecStart` で実行されるコマンドをターミナルで実行した場合、プロセスがバックグラウンドで実行されてコンソールがすぐ返ってきますか？

もしコンソールが返ってこないなら（プロセスがフォアグラウンドで動くなら）、`ExecStop` の指定は不要かもしれません。`ExecStop` を指定しない場合はデフォルトで（`KillSignal` を設定しない場合）`SIGTERM` が `ExecStart` で起動したプロセスに送られます。

> If this option is not specified, the process is terminated by sending the signal specified in KillSignal= or RestartKillSignal= when service stop is requested. 
>
> **[systemd.service](https://www.freedesktop.org/software/systemd/man/systemd.service.html#:~:text=If%20this%20option%20is%20not,service%20stop%20is%20requested.)**

もし起動したプロセスがシグナルをハンドリングしない場合は、`ExecStop` の指定が必要ですが、その場合は `package.json` に `"stop"` の定義が必要な気がします。npm のドキュメントを読むと、`"stop"` にはデフォルトで何も定義されていないようなので。

> Unlike with npm start, there is no default script that will run if the "stop" property is not defined.
>
> **[npm-stop | npm Docs](https://docs.npmjs.com/cli/v7/commands/npm-stop#:~:text=Unlike%20with,.)**

サンプルアプリ `v1.0.0` の `package.json` を確認すると、`"stop"` は定義されていませんでした。

> ```json
>   "scripts": {
>     "start": "json-server index.js --middlewares middleware.js"
>   },
> ```
> **[lerna-sample-payment-app/package.json at v1.0.0 · lerna-stack/lerna-sample-payment-app](https://github.com/lerna-stack/lerna-sample-payment-app/blob/v1.0.0/docker/mock-server/package.json#L6-L8)**

xirc replied:
@negokaz 

> ExecStart で実行されるコマンドをターミナルで実行した場合、プロセスがバックグラウンドで実行されてコンソールがすぐ返ってきますか？

ありがとうございます。確認しました。
プロセスはフォアグラウンドで実行されました。
停止するためには SIGTERM を送るほうが適切のようです。
ExecStop の記述を削除します。

xirc replied:
@negokaz 

ExecStop の記述を削除し、SIGTERM でプロセスを終了させるように変更しました。
> Mock server should be terminated by SIGTERM … 569bcbb

何度か ` sudo systemctl restart mock-server` を実行したのち、  
`ps` コマンドで複数のnpmプロセスが起動していないことも確認しました。
```shell
$ ps -aux | grep npm
root     18591  3.7  0.2 674436 31604 ?        Ssl  04:59   0:00 npm
ec2-user 18619  0.0  0.0 112816   976 pts/0    S+   04:59   0:00 grep --color=auto np
```

negokaz commented on examples/aws_ec2/facility-mock-server.tf:
```diff
@@ -37,88 +31,43 @@ resource "null_resource" "setup_for_mock_server" {
-    curl -L https://github.com/docker/compose/releases/download/1.28.5/docker-compose-`uname -s`-`uname -m` | sudo tee /usr/local/bin/docker-compose > /dev/null
-    sudo chmod +x /usr/local/bin/docker-compose
+    sudo --askpass install --owner=root --group=lv4 --mode=0664 /tmp/mock-server.service /usr/lib/systemd/system/mock-server.service
+    sudo systemctl restart mock-server
```
@xirc 

Systemd の設定ファイルを配置したあとは `systemctl daemon-reload` を実行したほうが良さそうです

> 新規のユニットファイルを作成したり、既存のユニットファイルを変更したら、systemctl daemon-reload コマンドを常に実行します。
>
> **[第10章 systemd によるサービス管理 Red Hat Enterprise Linux 7 | Red Hat Customer Portal](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/system_administrators_guide/chap-managing_services_with_systemd#:~:text=%E6%96%B0%E8%A6%8F%E3%81%AE,%E3%81%BE%E3%81%99%E3%80%82)**

xirc replied:
@negokaz 
systemd の 設定ファイルを更新したあとに、`systemctl daemon-reload` を実行するように変更しました。
> Run `systemctl daemon-reload` after updating the systemd unit file … 4f5651e

Commit:
Mock server should be terminated by SIGTERM

Remove `ExecStop` to terminate the server by SIGTERM.
https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStop=

Commit:
Run `systemctl daemon-reload` after updating the systemd unit file

This is required to avoid a systemctl execution failure.
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/system_administrators_guide/chap-managing_services_with_systemd#sect-Managing_Services_with_systemd-Unit_File_Create
